File: python/121_Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        return self.maxProfit2(prices)

    # compress space
    def maxProfit2(self, prices: List[int]) -> int:
        if not prices:
            return 0

        days = len(prices)
        if days == 0 or days == 1:
            return 0
        
        minn = prices[0]
        maxx = 0
        for i in range(days):
            maxx = max(maxx,prices[i]-minn)
            minn = min(minn,prices[i])
        return maxx
    
    def maxProfit1(self, prices: List[int]) -> int:
        if not prices:
            return 0

        days = len(prices)
        if days == 0 or days == 1:
            return 0
        dp = [[0] * 2 for _ in range(days+1)]
        dp[0][0] = 0 #max
        dp[0][1] = float('inf') #min

        for i in range(1,days+1):
            dp[i][0] = max(dp[i-1][0],prices[i-1]-dp[i-1][1])
            dp[i][1] = min(dp[i-1][1],prices[i-1])
        return dp[-1][0]

File: java/121_Best_Time_to_Buy_and_Sell_Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        return maxProfit1(prices);
    }

    public int maxProfit1(int[] prices) {
        int days = prices.length;
        if(days <= 1){
            return 0;
        }
        int[][] dp = new int[days+1][2];
        dp[0][0] = 0; // maxx
        dp[0][1] = prices[0]; // minn
        for(int i = 1 ; i < days + 1 ; i++){
            dp[i][0] = Math.max(dp[i-1][0] , prices[i-1] - dp[i-1][1]);
            dp[i][1] = Math.min(dp[i-1][1] , prices[i-1]);
        }
        return dp[days][0];

    }

    public int maxProfit2(int[] prices) {
        int maxx = 0;
        int minn = prices[0];
        for( int price : prices){
            maxx = Math.max(maxx,price-minn);
            minn = Math.min(minn,price);
        }
        return maxx;
    }

}